### PR TITLE
drivers: serial: Runtime Configure for SAM UART and USART

### PR DESCRIPTION
**Pull request by nick-kraus**
Implements dynamic configuration for the atmel,sam-uart and atmel,sam-usart drivers and enables the new API calls if CONFIG_UART_USE_RUNTIME_CONFIGURE is set.

Test samples [are available here](https://github.com/nick-kraus/zephyr/tree/add_sam_uart_configure-test/samples) (sam_{uart/usart}_configure_test), which checks that the expected configurations can be set and data can still be sent and received after dynamically changing the UART configuration.

Signed-off-by: Nick Kraus <nick@nckraus.com>

### CODE REVIEW

**Comment by nick-kraus** [comment]
It appears that the register header for the USART peripheral for the sam4l line of chips only has a different `#define` macro name than for all the other chip lines [(see here)](https://github.com/zephyrproject-rtos/hal_atmel/blob/master/asf/sam/include/sam4l/component/usart.h#L1004). For all other chips the symbol name is `US_MR_CHRL_{n}_BIT`, but for the 4l chip series it is `US_MR_CHRL_{n}`.

It is probably best to change the hal_atmel code to match, instead of using a #if here, so I'll work on a pull request for that as well, and this can be marked as draft or DNM until that is ready. It should however be ready to review, as the change in hal_atmel shouldn't require any changes here.

**Review by dcpleung — DISMISSED** [review]

**Comment by nick-kraus** [comment]
This pull request is now dependent on [a pull request on hal_atmel](https://github.com/zephyrproject-rtos/hal_atmel/pull/28), which defines some needed register masks and resolves the build failures for the sam4l line of chips, but the core of the PR wasn't changed in the last commit.

**Comment by nandojve** [comment]
Hi @nick-kraus ,

Please, update:
  * rebase on top of current main
  * update hash with latest from hal_atmel
  * reorder commits to add west changes first

Thank you!

**Comment by nick-kraus** [comment]
@nandojve I've performed the suggested changes, AFAIK this should be ready on my end to merge, just needs the re-reviews after the new push.

File: drivers/serial/usart_sam.c [hunk 1]
```diff
@@ -154,23 +96,213 @@ static int usart_sam_err_check(const struct device *dev)
...
+	/* Driver doesn't support 9 data bits, 0.5 stop bits, or DTR DSR flow control */
+	if (cfg->data_bits == UART_CFG_DATA_BITS_9 ||
+		cfg->stop_bits == UART_CFG_STOP_BITS_0_5 ||
+		cfg->flow_ctrl == UART_CFG_FLOW_CTRL_DTR_DSR) {
+		return -ENOTSUP;
+	}
+
+	/* Reset and disable USART */
+	usart->US_CR =   US_CR_RSTRX | US_CR_RSTTX
+		       | US_CR_RXDIS | US_CR_TXDIS | US_CR_RSTSTA;
```
Comment (nandojve) [inline drivers/serial/usart_sam.c:272]:
```
	usart->US_CR = US_CR_RSTRX | US_CR_RSTTX
		     | US_CR_RXDIS | US_CR_TXDIS
		     | US_CR_RSTSTA;
```

File: drivers/serial/usart_sam.c [hunk 2]
```diff
@@ -154,23 +96,213 @@ static int usart_sam_err_check(const struct device *dev)
...
+
+	/* normal UART mode, baud rate driven by peripheral clock, all
+	 * other values chosen by config
+	 */
+	usart->US_MR = US_MR_CHMODE_NORMAL
+		| US_MR_USCLKS_MCK
+		| usart_sam_cfg2sam_parity(cfg->parity)
+		| usart_sam_cfg2sam_stop_bits(cfg->stop_bits)
+		| usart_sam_cfg2sam_data_bits(cfg->data_bits)
+		| usart_sam_cfg2sam_flow_ctrl(cfg->flow_ctrl);
```
Comment (nandojve) [inline drivers/serial/usart_sam.c:282]:
```
	usart->US_MR = US_MR_CHMODE_NORMAL
		     | US_MR_USCLKS_MCK
		     | usart_sam_cfg2sam_parity(cfg->parity)
		     | usart_sam_cfg2sam_stop_bits(cfg->stop_bits)
		     | usart_sam_cfg2sam_data_bits(cfg->data_bits)
		     | usart_sam_cfg2sam_flow_ctrl(cfg->flow_ctrl);
```

File: drivers/serial/uart_sam.c [hunk 3]
```diff
@@ -145,23 +95,120 @@ static int uart_sam_err_check(const struct device *dev)
...
+	/* Driver only supports 8 data bits, 1 stop bit, and no flow control */
+	if (cfg->stop_bits != UART_CFG_STOP_BITS_1 ||
+		cfg->data_bits != UART_CFG_DATA_BITS_8 ||
+		cfg->flow_ctrl != UART_CFG_FLOW_CTRL_NONE) {
+		return -ENOTSUP;
+	}
+
+	/* Reset and disable UART */
+	uart->UART_CR =   UART_CR_RSTRX | UART_CR_RSTTX
+			| UART_CR_RXDIS | UART_CR_TXDIS | UART_CR_RSTSTA;
```
Comment (nandojve) [inline drivers/serial/uart_sam.c:181]:
```
	uart->UART_CR = UART_CR_RSTRX | UART_CR_RSTTX
		      | UART_CR_RXDIS | UART_CR_TXDIS
		      | UART_CR_RSTSTA;
```

File: drivers/serial/uart_sam.c [hunk 4]
```diff
@@ -145,23 +95,120 @@ static int uart_sam_err_check(const struct device *dev)
...
+
+	/* baud rate driven by the peripheral clock, UART does not filter
+	 * the receive line, parity chosen by config
+	 */
+	uart->UART_MR = UART_MR_CHMODE_NORMAL
+			| uart_sam_cfg2sam_parity(cfg->parity);
```
Comment (nandojve) [inline drivers/serial/uart_sam.c:187]:
```
	uart->UART_MR = UART_MR_CHMODE_NORMAL
		      | uart_sam_cfg2sam_parity(cfg->parity);
```

**Review by nandojve — CHANGES_REQUESTED** [review]
Hi @nick-kraus ,

Overall LGTM ! Please, fix that small alignments.

**Comment by nick-kraus** [comment]
> Hi @nick-kraus ,
> 
> I noted that your last update not aligned `=` with `|` yet. Use TABs most you can and then complete with spaces.

Apologies, just noticed that the zephyr repo has a tab width of 8, whereas my editor was set to 4. The last commit was to resolve that comment, but that obviously didn't work out. I'll make an update tonight with the proper width.

**Review by nandojve — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Pull request merged**